components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs

```diff
@@ -30,11 +30,56 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
         public int NumberOfConnectionRetries => _subscriptions?.FirstOrDefault()?.
             Subscription?.NumberOfConnectionRetries ?? 0;
 
+        private IDictionary<DateTime, int> _valueChanges = new Dictionary<DateTime, int>(100);
         /// <inheritdoc/>
-        public int ValueChangesCount { get; private set; } = 0;
+        public int ValueChangesCountLastMinute
+        {
+            get
+            {
+                return _valueChanges.Values.Sum();
+            }
+            private set
+            {
+                var difference = Math.Abs(value);
+                var now = DateTime.UtcNow;
+                var oneMinuteTimespan = TimeSpan.FromMinutes(1);
+                _valueChanges.Add(now, difference);
+
+                var valuesToDelete = _valueChanges.Where(kvp => now - kvp.Key > oneMinuteTimespan).Select(kvp => kvp.Key);
+                foreach (var valueToDelete in valuesToDelete) {
+                    _valueChanges.Remove(valueToDelete);
+                }
+            }
+        }
+
+        private int _valueChangesCount = 0;
+        /// <inheritdoc/>
+        public int ValueChangesCount { get { return _valueChangesCount; } private set { var difference = (value - _valueChangesCount);  _valueChangesCount = value; ValueChangesCountLastMinute = difference; } }
+
+        private IDictionary<DateTime, int> _dataChanges = new Dictionary<DateTime, int>(100);
+        /// <inheritdoc/>
+        public int DataChangesCountLastMinute {
+            get
+            {
+                return _dataChanges.Values.Sum();
+            }
+            private set
+            {
+                var difference = Math.Abs(value);
+                var now = DateTime.UtcNow;
+                var oneMinuteTimespan = TimeSpan.FromMinutes(1);
+                _dataChanges.Add(now, difference);
```

Comment:
Do we need to worry about the case when there is already a value for `now` entry ? I guess not, but I wanted to raise the question.

Reply:
The same question applies for `_valueChanges` on line `46`.

Reply:
changed implementation to ring buffer, not valid anymore